cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRuntimeLifecycleService.java

```diff
@@ -0,0 +1,364 @@
+import io.cdap.cdap.security.spi.authorization.UnauthorizedException;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import javax.annotation.Nullable;
+import org.apache.twill.api.RunId;
+import org.apache.twill.api.logging.LogEntry;
+
+public class ProgramRuntimeLifecycleService extends AbstractProgramLifecycleService {
+
+  private final ProgramRuntimeService runtimeService;
+
+  @Inject
+  public ProgramRuntimeLifecycleService(ProgramRuntimeService runtimeService, Store store,
+      AccessEnforcer accessEnforcer, AuthenticationContext authenticationContext,
+      PropertiesResolver propertiesResolver, ProgramStateWriter programStateWriter,
+      CapabilityReader capabilityReader, ArtifactRepository artifactRepository) {
+    super(store, programStateWriter, accessEnforcer, authenticationContext, propertiesResolver,
+        capabilityReader, artifactRepository);
+    this.runtimeService = runtimeService;
+  }
+
+  @Override
+  public void checkConcurrentExecution(ProgramId programId) throws Exception {
+    if (isConcurrentRunsInSameAppForbidden(programId.getType())) {
+      Map<RunId, RuntimeInfo> runs = runtimeService.list(programId);
+      if (!runs.isEmpty() || !isStoppedInSameProgram(programId)) {
+        throw new ConflictException(
+            String.format(
+                "Program %s is already running in an version of the same application with run ids %s",
+                programId, runs.keySet()));
+      }
+    }
+    if (!isConcurrentRunsAllowed(programId.getType())) {
+      List<RunId> runIds = new ArrayList<>();
+      for (Map.Entry<RunId, RuntimeInfo> entry : runtimeService.list(programId.getType())
+          .entrySet()) {
+        if (programId.equals(entry.getValue().getProgramId())) {
+          runIds.add(entry.getKey());
+        }
+      }
+      if (!runIds.isEmpty() || !isStopped(programId)) {
+        throw new ConflictException(
+            String.format("Program %s is already running with run ids %s", programId, runIds));
+      }
+    }
+  }
+
+  /**
+   * Starts a Program run with the given arguments. This method skips cluster lifecycle steps and
+   * does not perform authorization checks. If the program is already started, returns the
+   * controller for the program. NOTE: This method should only be used from this service and the
+   * {@link ProgramNotificationSubscriberService} upon receiving a {@link
+   * ProgramRunClusterStatus#PROVISIONED} state.
+   *
+   * @param programDescriptor descriptor of the program to run
+   * @param programOptions options for the program run
+   * @param programRunId program run id
+   * @return controller for the program
+   */
+  ProgramController startInternal(ProgramDescriptor programDescriptor,
```

Comment:
I would move it to {Abstract}ProgramRuntimeService. It is used outside of this class, also it uses only runtime service methods and seems more like a "sugar" over runtimeService.run. You can even remove the `run` from the interface as it's the only place where it's used and having it exposed makes this synchronized block look strange (as we leave a back door wide open)

Reply:
Overall, I'd probably reconsider the relations between this class and ProgramRuntimeService. It's not clear what are responsibilities of each, at least from the name. This guy does not do the "lifecycle" anymore, as it's mostly focused on starting and getting various runtime settings / statuses. Are there a value of having it separate from ProgramRuntimeService? At least as soon as we already use inheritance in the ProgramRuntimeService.

Reply:
Merged most of the code related to runtimeInfo to `ProgramRuntimeService` and common code to `ProgramUtil`.